src/style-spec/feature_filter/index.js

```diff
@@ -74,21 +74,161 @@ const filterSpec = {
+            dynamicFilter: (globalProperties: GlobalProperties, feature: Feature, canonical?: CanonicalTileID) => compiledDynamicFilter.value.evaluate(globalProperties, feature, {}, canonical),
+            needGeometry
+        };
+    }
+}
+
+function extractStaticFilter(filter: any): any {
+    if (!isDynamicFilter(filter)) {
+        return filter;
+    }
+
+    // Shallow copy so we can replace expressions in-place
+    let result = filter.slice();
+
+    // 1. Union branches
+    unionDynamicBranches(result);
+
+    // 2. Collapse dynamic conditions to  `true`
+    result = collapseDynamicBooleanExpressions(result);
+
+    return result;
+}
+
+function collapseDynamicBooleanExpressions(expression: any): any {
+    if (!Array.isArray(expression)) {
+        return expression;
+    }
+
+    const collapsed = collapsedExpression(expression);
+    if (collapsed === true) {
+        return collapsed;
+    } else {
+        return collapsed.map((subExpression) => collapseDynamicBooleanExpressions(subExpression));
+    }
+}
+
+const dynamicConditionExpressions = new Set([
+    'in',
+    '==',
+    '!=',
+    '>',
+    '>=',
+    '<',
+    '<=',
+    'to-boolean'
+]);
+
+function collapsedExpression(expression: any): any {
+    if (dynamicConditionExpressions.has(expression[0])) {
+        for(const param of expression.slice(1)) {
+            if (isDynamicFilter(param)) {
+                return true;
+            }
+        }
+    }
+    return expression;
+}
+
+/**
+ * Traverses the expression and replaces all instances of branching on a
+ * `dynamic` conditional (such as `['pitch']` or `['distance-from-center']`)
+ * into an `any` expression.
+ * This ensures that all possible outcomes of a `dynamic` branch are considered
+ * when evaluating the expression upfront during filtering.
+ *
+ * @param {Array<any>} filter the filter expression mutated in-place.
+ */
+function unionDynamicBranches(filter: any) {
+    const tests = [];
+    const branches = [];
+
+    if (filter[0] === 'case') {
+        for (let i = 1; i < filter.length - 1; i += 2) {
+            tests.push(filter[i]);
+            branches.push(filter[i + 1]);
+        }
+
+        branches.push(filter[filter.length - 1]);
+    } else if (filter[0] === 'match') {
+        tests.push(filter[1]);
+
+        for (let i = 2; i < filter.length - 1; i +=2) {
+            branches.push(filter[i + 1]);
+        }
+        branches.push(filter[filter.length - 1]);
+    } else if (filter[0] === 'step') {
+        tests.push(filter[1]);
+
+        for (let i = 1; i < filter.length - 1; i +=2) {
+            branches.push(filter[i + 1]);
+        }
     }
+
+    const isBranchingDynamically = tests.some((test) => isDynamicFilter(test));
+    if (isBranchingDynamically) {
+        filter.splice(0, filter.length);
+        filter.push('any');
+        filter.push(...branches);
```

Comment:
Nit: you can avoid allocating a new array for the return value of splice by setting `length`:

```suggestion
        filter.length = 0;
        filter.push('any', ...branches);
```